src/luxonis_ml/ops/artifacts.py

```diff
@@ -0,0 +1,267 @@
+                    if not local_dir.startswith(self.dataset.bough.value):
+                        local_dir = f"{self.dataset.bough.value}/{local_dir}"
+                    os.makedirs(f"tmp/{local_dir}", exist_ok=True)
+                    cmd = f"cp {relative_path} tmp/{local_dir}"
+                    subprocess.check_output(cmd, shell=True)
+
+                print("Uploading data to S3 - this could take some time...")
+                cmd = f"b2 sync tmp/{self.dataset.bough.value} b2://{self.dataset.bucket}/lakefs/{self.name}/_ldf_commits/{ldf_commit_number}/"
+                subprocess.check_output(cmd, shell=True)
+
+                folders = os.listdir(f"tmp/{self.dataset.bough.value}")
+                cmd = f"rm -rf tmp/"
+                subprocess.check_output(cmd, shell=True)
+
+                print("Staging in LakeFS...")
+                for path in tqdm(path_to_info):
+                    staged_object = self._stage_object(path, path_to_info)
+                    self.lakefs_client.objects.stage_object(self.name, self.dataset.branch, path, staged_object)
+
+                commit_creation = CommitCreation(message=message)
+                try:
+                    response = self.lakefs_client.commits.commit(self.name, self.dataset.branch, commit_creation)
+                    self.dataset.commit_id = response['id']
+                except:
+                    raise Exception("No commit - no changes!")
+
+            if len(self.dataset.removed_files[self.dataset.bough.value]) > 0:
+                print("Removing data")
+                for path in tqdm(self.dataset.removed_files[self.dataset.bough.value]):
+                    if path.startswith(self.dataset.bough.value):
+                        split_by = f"{self.dataset.bough.value}/"
+                    else:
+                        split_by = f"/{self.dataset.bough.value}/"
+                    lakefs_path = path.split(split_by)[-1]
+                    cmd = f"lakectl fs rm lakefs://{self.name}/{self.dataset.branch}/{lakefs_path}"
+                    subprocess.check_output(cmd, shell=True)
+
+                commit_creation = models.CommitCreation(message=message)
+                response = self.lakefs_client.commits.commit(self.name, self.dataset.branch, commit_creation)
+                self.dataset.commit_id = response['id']
+
+            self.dataset._reset_status()
+
+            t1 = time.time()
+            print(f"Took {(t1-t0)/60} minutes!")
+
+        else:
+            raise Exception("Must checkout a branch before pushing!")
+
+    def pull(self, s3_pull=False):
+
+        if hasattr(self.dataset, 'commit_id'):
+            t0 = time.time()
+
+            if s3_pull:
+                print("Running rclone to sync with S3...")
+                bucket_path = self.dataset.s3_path.split('s3://')[-1]
+                cmd = f"rclone sync --s3-force-path-style=true lakefs:{self.name}/{self.dataset.commit_id}/ b2:{bucket_path}/{self.dataset.bough.value}"
```

Comment:
Would it make sense to keep b2 as a variable somewhere in case we change the provider?

Reply:
This is now a variable in the `pull` function. Good call, as it depends on `rclone config`. It is also now an option in `luxonis_ml` CLI